rust-runtime/aws-smithy-observability/src/meter.rs

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Metrics are used to gain insight into the operational performance and health of a system in
+//! real time.
+
+use crate::attributes::{Attributes, Context};
+use crate::error::ObservabilityError;
+
+/// Provides named instances of [Meter].
+pub trait MeterProvider {
+    /// Get or create a named [Meter].
+    fn get_meter(&self, scope: &'static str, attributes: Option<&Attributes>) -> Box<dyn Meter>;
+
+    /// Optional method to flush the metrics pipeline, default is noop
+    fn flush(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+
+    /// Optional method to shutdown the metrics provider, default is noop
+    fn shutdown(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+}
+
+/// The entry point to creating instruments. A grouping of related metrics.
+pub trait Meter {
+    /// Create a new Gauge.
+    #[allow(clippy::type_complexity)]
+    fn create_gauge(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = f64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = f64>>;
```

Comment:
Same question should these be `Box<T> `or should they be `impl -> T`?

Reply:
See response above, using `impl T` would make these not object safe
https://github.com/smithy-lang/smithy-rs/pull/3912#discussion_r1841211316